### PR TITLE
HLRC: Create skeleton code for xpack APIs

### PR DESCRIPTION
**Pull request by hub-cap**

### CODE REVIEW

**Comment by hub-cap** [comment]
@spinscale just FYI i copy-pasta'd some of your changes into this PR so it can be merged ASAP. If yours merges first then ill gladly remove my changes that include the watcher client and changes to the tests/license.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/XPackClient.java [hunk 1]
```diff
@@ -40,11 +40,47 @@
...
+        this.licenseClient = new LicenseClient(restHighLevelClient);
+        this.migrationClient = new MigrationClient(restHighLevelClient);
+        this.mlClient = new MLClient(restHighLevelClient);
+        this.monitoringClient = new MonitoringClient(restHighLevelClient);
+        this.rollupClient = new RollupClient(restHighLevelClient);
+        this.securityClient = new SecurityClient(restHighLevelClient);
+        this.watcherClient = new WatcherClient(restHighLevelClient);
     }
 
+    public DeprecationClient getDeprecationClient() { return deprecationClient; }
```
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/XPackClient.java]: I think it is important that these have Javadoc and that it mention X-Pack.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client;
+
+public final class SecurityClient {
```
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java]: I think these should also have Javadoc and mention X-Pack.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MiscellaneousDocumentationIT.java [hunk 3]
```diff
@@ -97,8 +99,8 @@ public void testXPackInfo() throws Exception {
             //tag::x-pack-info-response
             BuildInfo build = response.getBuildInfo();                 // <1>
             LicenseInfo license = response.getLicenseInfo();           // <2>
-            assertEquals(XPackInfoResponse.BASIC_SELF_GENERATED_LICENSE_EXPIRATION_MILLIS,
-                    license.getExpiryDate());                          // <3>
+            assertThat(license.getExpiryDate(),
+                is(greaterThan(Instant.now().toEpochMilli())));        // <3>
```
Comment (nik9000) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MiscellaneousDocumentationIT.java]: Could you change the prose in the asciidoc file for this callout?

**Review by nik9000 — APPROVED** [review]
MOAR JAVADOC. Otherwise, LGTM.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client;
+
+public final class MLClient {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java]: not sure, but should this be MachineLearningClient?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MLIT.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client;
+
+public class MLIT extends ESRestHighLevelClientTestCase {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MLIT.java]: the name here seems a bit too concise as well :) Maybe MachineLearningIT ?

**Review by javanna — APPROVED** [review]
agreed on adding javadocs like Nik said, left a minor question on naming but LGTM. Thanks!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java (hunk 4 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java]: I considered this, as the IT is MLIT which is the nastiest name ever. But the spec shows them as `.ml.` so I wanted to stay true to it.. Do you consider renaming the class diverging, or is this ok? I could still do a `ml()` getter in the XPackClient so the `hlrc.xpack().ml()` would still work out.

Also, wow, i cant believe i did this... i need to remove the get's off those methods so they are not `hlrc.xpack().getML()` !!!!!

**Comment by hub-cap** [comment]
* Cleaned up the client getters to do only things like `.watcher()`, `.ml()`, etc. 
* Removed some classes which are not in the xpack completeness issue.
* javadoc'd all the things
* called out the trial and basic license info in the asciidoc.

@javanna or @nik9000 can you eyeball again?

**Review by nik9000 — APPROVED** [review]
LGTM

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java (hunk 4 above)
Comment (javanna) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/MLClient.java]: good point about the get prefix, I had missed it. I don't mind leaving naming as-is.

**Comment by hub-cap** [comment]
Closing as this info / code is out of date.

**Pull request closed without merging**